spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/test/AbstractProcessDeployerTests.java

```diff
@@ -0,0 +1,253 @@
+
+		String deploymentId = processDeployer().deploy(request);
+		Attempts timeout = deploymentTimeout();
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(deploying))), timeout.noAttempts, timeout.pause));
+
+		timeout = undeploymentTimeout();
+		processDeployer().undeploy(deploymentId);
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(unknown))), timeout.noAttempts, timeout.pause));
+
+	}
+
+	@Test
+	public void testFailedDeployment() {
+		Map<String, String> properties = new HashMap<>();
+		properties.put("killDelay", "0");
+		AppDefinition definition = new AppDefinition(randomName(), properties);
+		Resource resource = mavenResource("org.springframework.cloud.stream.module", "integration-test-processor","jar","exec","1.0.0.BUILD-SNAPSHOT");
+		AppDeploymentRequest request = new AppDeploymentRequest(definition, resource, properties);
+
+		String deploymentId = processDeployer().deploy(request);
+		Attempts timeout = deploymentTimeout();
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(failed))), timeout.noAttempts, timeout.pause));
+
+		timeout = undeploymentTimeout();
+		processDeployer().undeploy(deploymentId);
+		assertThat(deploymentId, eventually(hasStatusThat(
+				Matchers.<ProcessStatus>hasProperty("state", is(unknown))), timeout.noAttempts, timeout.pause));
+	}
+
+	protected String randomName() {
+		return UUID.randomUUID().toString();
+	}
+
+	/**
+	 * Return the timeout to use for repeatedly querying a module while it is being deployed.
+	 * Default value is one minute, being queried every 5 seconds.
+	 */
+	protected Attempts deploymentTimeout() {
+		return new Attempts(12, 5000);
+	}
+
+	/**
+	 * Return the timeout to use for repeatedly querying a module while it is being un-deployed.
+	 * Default value is one minute, being queried every 5 seconds.
+	 */
+	protected Attempts undeploymentTimeout() {
+		return new Attempts(20, 5000);
+	}
+
+	private Resource mavenResource(String groupId, String artifactId, String ext, String classifier, String version) {
+		// TODO: actually resolve using Aether,
+		// or use integration-test-processor in every test and bring at build time in src/test/resources?
+		File localRepository = new File(System.getProperty("user.home") + File.separator + ".m2" +
+				File.separator + "repository");
+		String path = groupId.replace(".", File.separator) + File.separator
+				+ artifactId + File.separator + version + File.separator
+				+ artifactId + "-" + version + "-" + classifier + "." + ext;
+		File jarFile = new File(localRepository, path);
+		return new FileSystemResource(jarFile);
+	}
+
+
+
+	/**
+	 * Represents a timeout for querying status, with repetitive queries until a certain number have been made.
+	 *
+	 * @author Eric Bottard
+	 */
+	protected static class Attempts {
```

Comment:
This class name feels off, and the fact that in the comment above and in the actual var names used in the test code it's always referred to as "timeout" is confusing. That said, it seems like more than "timeout" as well since it encompasses the max number of attempts and the pause between each.